internal/service/firewall/firewall.go

```diff
@@ -2,149 +2,187 @@ package firewall
-		ReadContext:   resourceFirewallRulesRead,
-		UpdateContext: resourceFirewallRulesUpdate,
-		DeleteContext: resourceFirewallRulesDelete,
-		Importer: &schema.ResourceImporter{
-			StateContext: schema.ImportStatePassthroughContext,
-		},
-		Schema: map[string]*schema.Schema{
-			"server_id": {
-				Type:        schema.TypeString,
-				Description: "The unique id of the server to be protected the firewall rules",
+var (
+	_ resource.Resource                = &firewallRulesResource{}
+	_ resource.ResourceWithConfigure   = &firewallRulesResource{}
+	_ resource.ResourceWithImportState = &firewallRulesResource{}
+)
+
+func NewFirewallRulesResource() resource.Resource {
+	return &firewallRulesResource{}
+}
+
+type firewallRulesResource struct {
+	client *service.Service
+}
+
+func (r *firewallRulesResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_firewall_rules"
+}
+
+// Configure adds the provider configured client to the resource.
+func (r *firewallRulesResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	r.client, resp.Diagnostics = utils.GetClientFromProviderData(req.ProviderData)
+}
+
+type firewallRulesModel struct {
+	ID           types.String        `tfsdk:"id"`
+	ServerID     types.String        `tfsdk:"server_id"`
+	FirewallRule []firewallRuleModel `tfsdk:"firewall_rule"`
+}
+
+type firewallRuleModel struct {
+	Action                  types.String `tfsdk:"action"`
+	Comment                 types.String `tfsdk:"comment"`
+	DestinationAddressStart types.String `tfsdk:"destination_address_start"`
+	DestinationAddressEnd   types.String `tfsdk:"destination_address_end"`
+	DestinationPortStart    types.String `tfsdk:"destination_port_start"`
+	DestinationPortEnd      types.String `tfsdk:"destination_port_end"`
+	Direction               types.String `tfsdk:"direction"`
+	Family                  types.String `tfsdk:"family"`
+	ICMPType                types.String `tfsdk:"icmp_type"`
+	Protocol                types.String `tfsdk:"protocol"`
+	SourceAddressStart      types.String `tfsdk:"source_address_start"`
+	SourceAddressEnd        types.String `tfsdk:"source_address_end"`
+	SourcePortStart         types.String `tfsdk:"source_port_start"`
+	SourcePortEnd           types.String `tfsdk:"source_port_end"`
+}
+
+func (r *firewallRulesResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "FirewallRuless can be used to connect multiple Private Networks. UpCloud Servers on any attached network can communicate directly with each other.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			"server_id": schema.StringAttribute{
 				Required:    true,
-				ForceNew:    true,
+				Description: "The unique id of the server to be protected the firewall rules",
```

Comment:
I think I have used UUID in the other new Id descriptions, but this is a bit different since this refers Id of another resource  🤔

```suggestion
				Description: "The UUID of the server to be protected the firewall rules",
```